[serve] Improve serve config docs (#40598)
ray-project/ray: doc/source/serve/configure-serve-deployment.md

# Configure Ray Serve deployments

> In this guide, you will learn about:

Reviewer: WDYT about adding some guidance on why you need to configure deployments?

# Configure Ray Serve deployments

> This guide walks through the parameters that are configurable for a Ray Serve deployment, as well as the different locations where you can specify the parameters. You can also refer to the [API reference](../serve/api/doc/ray.serve.deployment_decorator.rst) for the `@serve.deployment` decorator.

Reviewer 2: Could you reformat this as bullet points that explain what the user will learn? [The "Set Up FastAPI and HTTP" intro section](https://docs.ray.io/en/releases-2.7.1/serve/http-guide.html#set-up-fastapi-and-http) is a good example.

This helps users quickly gauge whether the doc is useful to them without needing to read a paragraph.

- `health_check_timeout_s` - Duration in seconds, that replicas wait for a health check method to return before considering it as failed. Defaults to 30s.
- `graceful_shutdown_wait_loop_s` - Duration that replicas wait until there is no more work to be done before shutting down. Defaults to 2s.
- `graceful_shutdown_timeout_s` - Duration to wait for a replica to gracefully shut down before being forcefully killed. Defaults to 20s.

## How to configure a deployment

> There are 2 ways to specify the parameters mentioned above. First, you can specify them through your application code. Second, you can specify them through the Serve Config file that is recommended for production.

Author: Makes sense, I added something similar in the config.md file!

### Specify parameters through application code

> In most cases where you want to specify deployment parameters through your application code, you can use the `@serve.deployment` decorator:

Reviewer 2: ```suggestion
Use the `@serve.deployment` decorator to specify deployment parameters:
```

```{literalinclude} ../serve/doc_code/configure_serve_deployment/model_deployment.py
:start-after: __deployment_start__
:end-before: __deployment_end__
:language: python
```

> Sometimes, you might want to reuse an already-defined deployment but deploy it with different configurations; or you might want to dynamically set the parameters at runtime. In such cases, you can use `.options()` on a deployment to set or override parameters specified in the `@serve.deployment` decorator:

Reviewer 2: ```suggestion
Use the `.options()` method to modify deployment parameters on an already-defined deployment. This lets you reuse deployment definitions and dynamically set parameters at runtime.
```

```{literalinclude} ../serve/doc_code/configure_serve_deployment/model_deployment.py
:start-after: __deployment_end__
:end-before: __options_end__
:language: python
```

### Specify parameters through Serve config file

> Lastly, you can configure individual deployments in each application through the Serve config file. The recommended way to deploy and update your applications in production is through the Serve Config file. Learn more about how to use the Serve Config in the [production guide](https://docs.ray.io/en/latest/serve/production-guide/config.html).

Reviewer 2: ```suggestion
In production, we recommend configuring individual deployments through the Serve config file. Learn more about how to use the Serve Config in the [production guide](https://docs.ray.io/en/latest/serve/production-guide/config.html).
```

> Lastly, you can configure individual deployments in each application through the Serve config file. The recommended way to deploy and update your applications in production is through the Serve Config file. Learn more about how to use the Serve Config in the [production guide](https://docs.ray.io/en/latest/serve/production-guide/config.html).

Reviewer 2: Can you link this with a documentation reference instead of a URL?

- `health_check_timeout_s` - Duration in seconds, that replicas wait for a health check method to return before considering it as failed. Defaults to 30s.
- `graceful_shutdown_wait_loop_s` - Duration that replicas wait until there is no more work to be done before shutting down. Defaults to 2s.
- `graceful_shutdown_timeout_s` - Duration to wait for a replica to gracefully shut down before being forcefully killed. Defaults to 20s.

## How to configure a deployment

There are 2 ways to specify the parameters mentioned above. First, you can specify them through your application code. Second, you can specify them through the Serve Config file that is recommended for production.

### Specify parameters through application code

Use the `@serve.deployment` decorator to specify deployment parameters:

```{literalinclude} ../serve/doc_code/configure_serve_deployment/model_deployment.py
:start-after: __deployment_start__
:end-before: __deployment_end__
:language: python
```

> Use the `.options()` method to modify deployment parameters on an already-defined deployment. This lets you reuse deployment definitions and dynamically set parameters at runtime.

Reviewer: Link to the API ref for `.options()`?

    ray_actor_options:
      num_cpus: 0.2
      num_gpus: 0.0
```

### Overriding deployment settings

For each individual parameter, the order of priority is (from highest to lowest):

1. Serve Config file
2. Application code (either through the `@serve.deployment` decorator or through `.options()`)
3. Serve defaults

In other words, if a parameter for a deployment is specified in the config file and the application code, Serve will use the config file's value. If it's only specified in the code, Serve will use the value specified in the code. If the parameter is specified anywhere, Serve will use the default for that parameter.

For example, if you have the following application code which contains a single deployment `ExampleDeployment`:

```python
@serve.deployment(num_replicas=2, graceful_shutdown_timeout_s=6)
class ExampleDeployment:
    ...

example_app = ExampleDeployment.bind()
```

Then suppose you deploy your application with the following config file:

```yaml
applications:
  - name: default
    import_path: models:example_app 
    deployments:
      - name: ExampleDeployment
        num_replicas: 5
```

> Serve uses `num_replicas=5` from the value set in the config file and `graceful_shutdown_timeout_s=6` from the value set in the application code. All other deployment settings use Serve defaults because the user didn't specify them in the code or the config. For instance, `health_check_period_s=10` since by default Serve health checks deployments once every 10 seconds.

Reviewer 2: ```suggestion
Serve uses `num_replicas=5` from the value set in the config file and `graceful_shutdown_timeout_s=6` from the value set in the application code. All other deployment settings use Serve defaults because they're not specified in the code or the config. For instance, `health_check_period_s=10` since by default Serve health checks deployments once every 10 seconds.
```
